File: packages/edge/src/services/requestStats.ts
/**
 * Request Statistics Service
 * Collects and provides request statistics for monitoring
 * - Status code counts
 * - Endpoint counts
 * - Response time statistics (min, max, avg, p95, p99)
 * - Bytes sent/received totals
 * - Rate-limited request logging
 */

export interface EndpointStats {
  count: number;
  totalDurationMs: number;
  minDurationMs: number;
  maxDurationMs: number;
  durations: number[]; // For percentile calculations (limited size)
  bytesSent: number;
  bytesReceived: number;
  statusCodes: Map<number, number>; // Status codes per endpoint
}

export interface RequestStatsSnapshot {
  startTime: string;
  snapshotTime: string;
  uptimeSeconds: number;
  totalRequests: number;
  statusCodes: Record<string, number>;
  endpoints: Record<
    string,
    {
      count: number;
      avgDurationMs: number;
      minDurationMs: number;
      maxDurationMs: number;
      p95DurationMs: number;
      p99DurationMs: number;
      bytesSent: number;
      bytesReceived: number;
      statusCodes: Record<string, number>; // Per-endpoint status codes
    }
  >;
  totals: {
    bytesSent: number;
    bytesReceived: number;
    avgDurationMs: number;
    minDurationMs: number;
    maxDurationMs: number;
  };
}

class RequestStats {
  private startTime: Date;
  private totalRequests: number = 0;
  private statusCodes: Map<number, number> = new Map();
  private endpoints: Map<string, EndpointStats> = new Map();
  private totalBytesSent: number = 0;
  private totalBytesReceived: number = 0;
  private allDurations: number[] = [];
  private minDuration: number = Infinity;
  private maxDuration: number = 0;

  // Rate limiting for logging
  private logRateLimit: number; // requests per second
  private logTokens: number;
  private lastTokenRefill: number;

  // Max durations to keep for percentile calculations per endpoint
  private readonly maxDurationsPerEndpoint = 1000;
  private readonly maxTotalDurations = 10000;

  constructor() {
    this.startTime = new Date();
    // Default: 100 logs per second, 0 = disabled
    this.logRateLimit = parseInt(process.env.REQUEST_LOG_RATE_LIMIT || '100', 10);
    this.logTokens = this.logRateLimit;
    this.lastTokenRefill = Date.now();
  }

  /**
   * Record a request
   * @returns true if this request should be logged (rate limit check)
   */
  record(
    method: string,
    path: string,
    statusCode: number,
    durationMs: number,
    bytesSent: number,
    bytesReceived: number
  ): boolean {
    this.totalRequests++;

    // Status code counts
    const currentStatus = this.statusCodes.get(statusCode) || 0;
    this.statusCodes.set(statusCode, currentStatus + 1);

    // Endpoint stats (method + path)
    const endpointKey = `${method} ${path}`;
    let endpoint = this.endpoints.get(endpointKey);
    if (!endpoint) {
      endpoint = {
        count: 0,
        totalDurationMs: 0,
        minDurationMs: Infinity,
        maxDurationMs: 0,
        durations: [],
        bytesSent: 0,
        bytesReceived: 0,
        statusCodes: new Map(),
      };
      this.endpoints.set(endpointKey, endpoint);
    }

    endpoint.count++;
    endpoint.totalDurationMs += durationMs;
    endpoint.minDurationMs = Math.min(endpoint.minDurationMs, durationMs);
    endpoint.maxDurationMs = Math.max(endpoint.maxDurationMs, durationMs);
    endpoint.bytesSent += bytesSent;
    endpoint.bytesReceived += bytesReceived;

    // Track status code per endpoint
    const currentEndpointStatus = endpoint.statusCodes.get(statusCode) || 0;
    endpoint.statusCodes.set(statusCode, currentEndpointStatus + 1);

    // Keep limited durations for percentile calculation
    if (endpoint.durations.length < this.maxDurationsPerEndpoint) {
      endpoint.durations.push(durationMs);
    } else {
      // Replace random element to maintain sample diversity
      const idx = Math.floor(Math.random() * this.maxDurationsPerEndpoint);
      endpoint.durations[idx] = durationMs;
    }

    // Global stats
    this.totalBytesSent += bytesSent;
    this.totalBytesReceived += bytesReceived;
    this.minDuration = Math.min(this.minDuration, durationMs);
    this.maxDuration = Math.max(this.maxDuration, durationMs);

    // Keep limited total durations
    if (this.allDurations.length < this.maxTotalDurations) {
      this.allDurations.push(durationMs);
    } else {
      const idx = Math.floor(Math.random() * this.maxTotalDurations);
      this.allDurations[idx] = durationMs;
    }

    // Rate limit check for logging
    return this.shouldLog();
  }

  /**
   * Check if request should be logged based on rate limit
   */
  private shouldLog(): boolean {
    if (this.logRateLimit === 0) {
      return false; // Logging disabled
    }

    const now = Date.now();
    const elapsed = now - this.lastTokenRefill;

    // Refill tokens based on elapsed time (token bucket algorithm)
    if (elapsed >= 1000) {
      this.logTokens = this.logRateLimit;
      this.lastTokenRefill = now;
    } else {
      const tokensToAdd = (elapsed / 1000) * this.logRateLimit;
      this.logTokens = Math.min(this.logRateLimit, this.logTokens + tokensToAdd);
      this.lastTokenRefill = now;
    }

    if (this.logTokens >= 1) {
      this.logTokens--;
      return true;
    }

    return false;
  }

  /**
   * Calculate percentile from sorted array
   */
  private percentile(sorted: number[], p: number): number {
    if (sorted.length === 0) return 0;
    const idx = Math.ceil((p / 100) * sorted.length) - 1;
    return sorted[Math.max(0, idx)];
  }

  /**
   * Get statistics snapshot
   */
  getSnapshot(): RequestStatsSnapshot {
    const now = new Date();
    const uptimeSeconds = Math.floor((now.getTime() - this.startTime.getTime()) / 1000);

    // Status codes as object
    const statusCodesObj: Record<string, number> = {};
    for (const [code, count] of this.statusCodes) {
      statusCodesObj[code.toString()] = count;
    }

    // Endpoint stats
    const endpointsObj: Record<string, RequestStatsSnapshot['endpoints'][string]> = {};
    for (const [key, stats] of this.endpoints) {
      const sorted = [...stats.durations].sort((a, b) => a - b);
      // Convert endpoint status codes Map to object
      const endpointStatusCodes: Record<string, number> = {};
      for (const [code, count] of stats.statusCodes) {
        endpointStatusCodes[code.toString()] = count;
      }
      endpointsObj[key] = {
        count: stats.count,
        avgDurationMs: stats.count > 0 ? Math.round(stats.totalDurationMs / stats.count) : 0,
        minDurationMs: stats.minDurationMs === Infinity ? 0 : stats.minDurationMs,
        maxDurationMs: stats.maxDurationMs,
        p95DurationMs: this.percentile(sorted, 95),
        p99DurationMs: this.percentile(sorted, 99),
        bytesSent: stats.bytesSent,
        bytesReceived: stats.bytesReceived,
        statusCodes: endpointStatusCodes,
      };
    }

    // Calculate global average
    const totalDuration = Array.from(this.endpoints.values()).reduce(
      (sum, e) => sum + e.totalDurationMs,
      0
    );
    const avgDuration = this.totalRequests > 0 ? Math.round(totalDuration / this.totalRequests) : 0;

    return {
      startTime: this.startTime.toISOString(),
      snapshotTime: now.toISOString(),
      uptimeSeconds,
      totalRequests: this.totalRequests,
      statusCodes: statusCodesObj,
      endpoints: endpointsObj,
      totals: {
        bytesSent: this.totalBytesSent,
        bytesReceived: this.totalBytesReceived,
        avgDurationMs: avgDuration,
        minDurationMs: this.minDuration === Infinity ? 0 : this.minDuration,
        maxDurationMs: this.maxDuration,
      },
    };
  }

  /**
   * Reset all statistics
   */
  reset(): void {
    this.startTime = new Date();
    this.totalRequests = 0;
    this.statusCodes.clear();
    this.endpoints.clear();
    this.totalBytesSent = 0;
    this.totalBytesReceived = 0;
    this.allDurations = [];
    this.minDuration = Infinity;
    this.maxDuration = 0;
  }

  /**
   * Get current rate limit setting
   */
  getRateLimit(): number {
    return this.logRateLimit;
  }

  /**
   * Update rate limit at runtime
   */
  setRateLimit(limit: number): void {
    this.logRateLimit = limit;
    this.logTokens = limit;
  }
}

// Singleton instance
export const requestStats = new RequestStats();

File: packages/edge-dotnet/src/Gatrix.Edge/Services/RequestStats.cs
using System.Collections.Concurrent;

namespace Gatrix.Edge.Services;

/// <summary>
/// Collects and provides request statistics for monitoring.
/// Includes status code counts, endpoint counts, response time stats, and rate-limited logging.
/// </summary>
public class RequestStats
{
    private readonly DateTime _startTime = DateTime.UtcNow;
    private long _totalRequests;
    private readonly ConcurrentDictionary<int, long> _statusCodes = new();
    private readonly ConcurrentDictionary<string, EndpointStats> _endpoints = new();
    private long _totalBytesSent;
    private long _totalBytesReceived;
    private double _minDuration = double.MaxValue;
    private double _maxDuration;

    // Rate limiting
    private int _logRateLimit;
    private double _logTokens;
    private long _lastTokenRefill;
    private readonly object _rateLock = new();

    private const int MaxDurationsPerEndpoint = 1000;

    public RequestStats(int rateLimit = 100)
    {
        _logRateLimit = rateLimit;
        _logTokens = rateLimit;
        _lastTokenRefill = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
    }

    /// <summary>
    /// Record a request and return whether it should be logged (rate limited).
    /// </summary>
    public bool Record(string method, string path, int statusCode, double durationMs, long bytesSent, long bytesReceived)
    {
        Interlocked.Increment(ref _totalRequests);
        _statusCodes.AddOrUpdate(statusCode, 1, (_, c) => c + 1);

        Interlocked.Add(ref _totalBytesSent, bytesSent);
        Interlocked.Add(ref _totalBytesReceived, bytesReceived);

        // Min/Max (approximation for concurrent access)
        double oldMin, newMin;
        do { oldMin = _minDuration; newMin = Math.Min(oldMin, durationMs); }
        while (oldMin != newMin && Interlocked.CompareExchange(ref _minDuration, newMin, oldMin) != oldMin);

        double oldMax, newMax;
        do { oldMax = _maxDuration; newMax = Math.Max(oldMax, durationMs); }
        while (oldMax != newMax && Interlocked.CompareExchange(ref _maxDuration, newMax, oldMax) != oldMax);

        var endpointKey = $"{method} {path}";
        var endpoint = _endpoints.GetOrAdd(endpointKey, _ => new EndpointStats());
        endpoint.Record(statusCode, durationMs, bytesSent, bytesReceived);

        return ShouldLog();
    }

    private bool ShouldLog()
    {
        if (_logRateLimit == 0) return false;

        lock (_rateLock)
        {
            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var elapsed = now - _lastTokenRefill;

            if (elapsed >= 1000)
            {
                _logTokens = _logRateLimit;
                _lastTokenRefill = now;
            }
            else
            {
                var tokensToAdd = (elapsed / 1000.0) * _logRateLimit;
                _logTokens = Math.Min(_logRateLimit, _logTokens + tokensToAdd);
                _lastTokenRefill = now;
            }

            if (_logTokens >= 1)
            {
                _logTokens--;
                return true;
            }
            return false;
        }
    }

    public object GetSnapshot()
    {
        var now = DateTime.UtcNow;
        var uptimeSeconds = (int)(now - _startTime).TotalSeconds;

        var statusCodesObj = new Dictionary<string, long>();
        foreach (var kvp in _statusCodes)
            statusCodesObj[kvp.Key.ToString()] = kvp.Value;

        var endpointsObj = new Dictionary<string, object>();
        foreach (var kvp in _endpoints)
            endpointsObj[kvp.Key] = kvp.Value.GetSnapshot();

        var totalDuration = _endpoints.Values.Sum(e => e.TotalDurationMs);

        return new
        {
            startTime = _startTime.ToString("o"),
            snapshotTime = now.ToString("o"),
            uptimeSeconds,
            totalRequests = Interlocked.Read(ref _totalRequests),
            statusCodes = statusCodesObj,
            endpoints = endpointsObj,
            totals = new
            {
                bytesSent = Interlocked.Read(ref _totalBytesSent),
                bytesReceived = Interlocked.Read(ref _totalBytesReceived),
                avgDurationMs = _totalRequests > 0
                    ? Math.Round(totalDuration / Interlocked.Read(ref _totalRequests))
                    : 0,
                minDurationMs = _minDuration == double.MaxValue ? 0 : Math.Round(_minDuration),
                maxDurationMs = Math.Round(_maxDuration),
            },
        };
    }

    public void Reset()
    {
        _totalRequests = 0;
        _statusCodes.Clear();
        _endpoints.Clear();
        _totalBytesSent = 0;
        _totalBytesReceived = 0;
        _minDuration = double.MaxValue;
        _maxDuration = 0;
    }

    public int GetRateLimit() => _logRateLimit;
    public void SetRateLimit(int limit)
    {
        _logRateLimit = limit;
        _logTokens = limit;
    }

    private class EndpointStats
    {
        private long _count;
        private double _totalDurationMs;
        private double _minDurationMs = double.MaxValue;
        private double _maxDurationMs;
        private readonly List<double> _durations = new();
        private long _bytesSent;
        private long _bytesReceived;
        private readonly ConcurrentDictionary<int, long> _statusCodes = new();
        private readonly object _lock = new();

        public double TotalDurationMs => _totalDurationMs;

        public void Record(int statusCode, double durationMs, long bytesSent, long bytesReceived)
        {
            Interlocked.Increment(ref _count);
            Interlocked.Add(ref _bytesSent, bytesSent);
            Interlocked.Add(ref _bytesReceived, bytesReceived);
            _statusCodes.AddOrUpdate(statusCode, 1, (_, c) => c + 1);

            lock (_lock)
            {
                _totalDurationMs += durationMs;
                _minDurationMs = Math.Min(_minDurationMs, durationMs);
                _maxDurationMs = Math.Max(_maxDurationMs, durationMs);

                if (_durations.Count < MaxDurationsPerEndpoint)
                    _durations.Add(durationMs);
                else
                    _durations[Random.Shared.Next(MaxDurationsPerEndpoint)] = durationMs;
            }
        }

        public object GetSnapshot()
        {
            double[] sorted;
            lock (_lock)
            {
                sorted = _durations.OrderBy(d => d).ToArray();
            }

            var statusCodesObj = new Dictionary<string, long>();
            foreach (var kvp in _statusCodes)
                statusCodesObj[kvp.Key.ToString()] = kvp.Value;

            var count = Interlocked.Read(ref _count);
            return new
            {
                count,
                avgDurationMs = count > 0 ? Math.Round(_totalDurationMs / count) : 0,
                minDurationMs = _minDurationMs == double.MaxValue ? 0 : Math.Round(_minDurationMs),
                maxDurationMs = Math.Round(_maxDurationMs),
                p95DurationMs = Percentile(sorted, 95),
                p99DurationMs = Percentile(sorted, 99),
                bytesSent = Interlocked.Read(ref _bytesSent),
                bytesReceived = Interlocked.Read(ref _bytesReceived),
                statusCodes = statusCodesObj,
            };
        }

        private static double Percentile(double[] sorted, int p)
        {
            if (sorted.Length == 0) return 0;
            var idx = Math.Max(0, (int)Math.Ceiling(p / 100.0 * sorted.Length) - 1);
            return Math.Round(sorted[idx]);
        }
    }
}
